### ISSUE TITLE
Inheriting from generic class/struct ignoring the type causes segmentation fault

### ISSUE THREAD

**Issue by porras** [crystal-lang/crystal#1944]
I'm guessing that is not allowed (although I run into a case where it would be useful), but should probably give a compilation error:

Code:

```crystal
struct Generic(T)
end

struct IgnoreType < Generic(T)
end

p IgnoreType.new.is_a?(Generic(String))
p IgnoreType.new.is_a?(Generic(Int32))
```

```
$ crystal example.cr 
Segmentation fault: 11
```

The same happens if they are classes instead of structs.

(if this is allowed, I'd expect both lines to print `true`, and if it's not, a compilation error)

**Comment by lbguilherme** [comment]
This code as-is just doesn

**Comment by lbguilherme** [comment]
This code as-is just doesn't make much sense. Consider this:

```crystal
struct Generic(T)
  def foo
   T.new
  end
end

struct IgnoreType < Generic(T)
end

p IgnoreType.new.foo # What is this? (segfaults)
```

This should fail compilation with a "Unknown identifier T".
I'm curious, where would this be useful and what would be the semantics of it?

**Comment by asterite** [comment]
Hm, that should give a compile error right at the definition of `IgnoreType`.

**Comment by porras** [comment]
Of course that code only makes sense if no method that uses `T` is inherited.

A possible use case would be trying to define an `Option` type similar to Scala's. A possible implementation could be:

```crystal
abstract struct Option(T)
end

struct Some(T) < Option(T)
  def initialize(@value : T)
  end
end

struct None < Option(T)
end
```

Now, I'm not saying that this should work, I could totally imagine this not being possible (and in fact this kind of programming is not very Crystal-idiomatic I think), but then it should be a compilation error.

An important detail that I forgot to add is that the segmentation fault happens in the compiler, that's why I thought it is a bug worth reporting:

```
$ crystal build example.cr 
Segmentation fault: 11
```

**Comment by Perelandric** [comment]
Seems fixed as of v16.

>Error in line 4: can't extend non-abstract struct Generic(T)

**Comment by asterite** [comment]
I'm sure it fails with a class though

**Comment by Perelandric** [comment]
Indeed it does.

**Comment by asterite** [comment]
Closed in favor of #2665
